Use Botorch MultiTaskGP for transfer learning

AVHopp opened the pull request:
Replaces the custom `IndexKernel` construction with BoTorch's `MultiTaskGP` (which became possible due the added `all_tasks` argument).

Hrovatin commented:
Update on implementation:

- Created Model Factory logic similar to Kernel Factory. 
  - In contrast to the discussion with @AdrianSosic I did not create inheritance or other type of logic connecting the two as the only thing they share are being callable. So I do not see benefit of having shared logic.
- Added BoTorch preset model factory that creates BoTorch Model with default parameters and `SingleTaskGP` or `MultiTaskGP`
  - This model uses all possible defaults (outcome_transform, mean_module, covar_module, likelihood; except input_transform which is needed due to bound constraints). This is in contrast to what we discussed regarding step wise changes of all components. However, if we want to make a preset with defaults, I think this is a more appropriate course of action: 1.) We see if defaults perform well enough, 2.) if not explore why, 3.) if they perform better than current default, a separate investigation unrelated to this PR should be made to add potential improvements to current default.
  - The preset allows for overwriting of default BoTorch components.
  - Some changes in `GaussianProcessSurrogate` were required to satisfy the above logic.

Hrovatin commented:
Just to keep track of things that were attempted here (and done wrongly):
- Use BotorchKernelFactory which uses same parameters as BoTorch: https://github.com/emdgroup/baybe/pull/483
- Hard code using BoTorch model and defaults: https://github.com/emdgroup/baybe/blob/feature/botorch_multitaskgp/baybe/surrogates/gaussian_process/core.py and subsequent branches that hard-coded using botorch defaults https://github.com/emdgroup/baybe/blob/tl_kernel_preset/baybe/surrogates/gaussian_process/core.py, https://github.com/emdgroup/baybe/blob/tl_likelihood_preset/baybe/surrogates/gaussian_process/core.py, https://github.com/emdgroup/baybe/blob/tl_transform_preset/baybe/surrogates/gaussian_process/core.py
- Make botorch model preset: https://github.com/emdgroup/baybe/blob/tl_benchmarking_investigation_modelPreset/baybe/surrogates/gaussian_process/presets/botorch.py

Hrovatin committed:
Replace SingleTaskGP+IndexKernel with MultiTaskGP for transfer learning

Hrovatin committed:
Expand tests with different types of available data

Hrovatin commented:
Update on implementation:
- Starting from main, change only things that are strictly necessary for switching from SingleTaskGP to MultiTaskGP for transfer learning (this branch)
- Add StratifiedOutTransform as that is recommended by BoTorch and was before observed to improve performance. For now just added where model is selected. (https://github.com/emdgroup/baybe/blob/b5e6dce0ce5266cd66fd4fc0b99fb3e0be4ef4ab/baybe/surrogates/gaussian_process/core.py#L230)

AdrianSosic reviewed (commented):
Hi @Hrovatin, here the first batch of comments

Hrovatin committed:
Improve the approach used to get the single task parameter

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Hrovatin committed:
Correct anti-pattern in test

Co-authored-by: AdrianSosic <adrian.sosic@merckgroup.com>

Hrovatin commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -171,37 +172,63 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
             outcome_transform=outcome_transform,
             mean_module=mean_module,
-            covar_module=covar_module,
+            covar_module=base_covar_module,
```
renamed

Hrovatin commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -171,37 +172,63 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+                ],
+                "rank": context.n_tasks,
+                "task_covar_prior": None,
+                "all_tasks": task_param.comp_df[task_param.name].astype(int).to_list(),
```
Regarding 2nd point: haven't tested so far but it seems someone made sure there is a float here so I would imagine that this would cause error (maybe in torch?) otherwise https://github.com/emdgroup/baybe/blob/a56d418d88480052d1c72ec5fc1f137711f66278/baybe/parameters/categorical.py#L74

Hrovatin committed:
Implement review comments

Hrovatin commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -171,37 +172,63 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+                for p in context.searchspace.discrete.parameters
+                if isinstance(p, TaskParameter)
+            ][0]
+            if len(task_param.active_values) > 1:
```
Plus it would also not work for StratifiedOutputTransform as far as I remember: https://github.com/pytorch/botorch/issues/2739

AdrianSosic replied:
Yes, that's the current symptoms that are exposed code-wise. But I was more asking conceptually: What would we ideally expect to happen here? Currently, I see no reason that speaks against having a user request a recommendation for a searchspace with two active tasks, and I wonder what would need to change in our machinery to enable that. Not saying that we need to do it right now – might be better suited when enabling multi-fidelity. But any idea already?

Hrovatin replied:
This was resolved by Target refactoring. The scaling would still be an issue, but we currently do not use it so closing this PR

Hrovatin committed:
Clarify why single active value is required

Hrovatin committed:
Improve test description

Hrovatin committed:
Change TaskParameter computational representation to int

Hrovatin committed:
Remove int conversion for new int-based comp_df of task parameter

Scienfitz commented:
@Hrovatin any update?

AVHopp reviewed (commented):
First round of comments, but we should discuss some of the points (in particular the one regarding multiple active values) internally first.

Hrovatin committed:
Add test for transfer learning with multiple active task parameter values

Hrovatin committed:
Remove constraint to use single active task parameter value

Hrovatin committed:
Update tests and assert that multiple active values are recommended

Hrovatin commented on baybe/parameters/categorical.py:
```diff
@@ -87,6 +87,20 @@ class TaskParameter(CategoricalParameter):
 
+    @override
+    @cached_property
+    def comp_df(self) -> pd.DataFrame:
```
MultiTaskGP requires some values to be int not float hence we force conversion.

Scienfitz replied:
if I understood @AdrianSosic correctly (as part of discussions around #671 ) it doesnt matter if we take care of the dtype here as internally there is a cast to int and all tensors must be passed as floats or so, so tldr seems this change is obsolete. Reopening here just to make sure it was seen and removed / adapted properly

Hrovatin replied:
Currently we dont do it internally exactly because @AdrianSosic suggested overriding
(The code within the GP class is `task_comp_rep = task_param.comp_df.iloc[:, 0]`)

Hrovatin committed:
Remove mypy errors

Hrovatin committed:
Remove check that both tasks were recommended as this may not always happen

AVHopp commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -167,41 +168,67 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+                for p in context.searchspace.discrete.parameters
+                if isinstance(p, TaskParameter)
+            )
+            task_comp_rep = task_param.comp_df.iloc[
+                :, 0
+            ]  # using iloc instead of squeeze to avoid mypy errors
```
Avoid weird line break. Note that this has been suggested via Github interface and might thus not be 100% properly formatted.

AVHopp commented on tests/test_transfer_learning.py:
```diff
@@ -0,0 +1,85 @@
+    )
+
+    campaign.add_measurements(lookup)
+
```
```suggestion
```

AVHopp commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -167,41 +168,67 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+            model_kwargs = {}
+        else:
+            model_cls = botorch.models.MultiTaskGP
+            # TODO - is te below ok?
+            #  It is required that there is only one task parameter with only
+            #  one active value.
+            #  ...
+            #  One active task value is required for MultiTaskGP as else
+            #  one posterior per task would be returned:
+            #  https://github.com/pytorch/botorch/blob/a018a5ffbcbface6229d6c39f7ac6ef9baf5765e/botorch/models/gpytorch.py#L951
+            #  This would cause issues with current acquisition function implementation.
+            #  ...
```
This comment does no longer apply, right? So can it just be deleted or is there anything that is still open/problematic here?

AVHopp reviewed (commented):
Would be willing to approve - however, since this is technically my PR I can't

Hrovatin commented:
Results after rebase:
Note:
- Hartmann tl did not run for the new branch due to issues in local setup (shows only the main branch). But I tested that it runs successfully in actions
- Reproducibility is in general not 100% (also when not using the tl code that was changed)
<img width="716" height="3490" alt="image" src="https://github.com/user-attachments/assets/3f6de38f-5843-4049-91fa-51e5d4ed3c06" />

Hrovatin committed:
Update baybe/surrogates/gaussian_process/core.py

Co-authored-by: Alexander V. Hopp <alexander.hopp@merckgroup.com>

Hrovatin committed:
Update tests/test_transfer_learning.py

Co-authored-by: Alexander V. Hopp <alexander.hopp@merckgroup.com>

Hrovatin commented on .github/workflows/benchmark.yml:
```diff
@@ -17,12 +17,14 @@ on:
           - "Transfer Learning"
           - "Synthetic"
           - "Non Transfer Learning"
+          - "TL Hartmann"
```
```suggestion
```

Hrovatin committed:
Remove unnecessary comments

Hrovatin commented on tests/test_transfer_learning.py:
```diff
@@ -0,0 +1,85 @@
+        objective=objective,
+    )
+
+    if lookup.shape[0] > 0:
```
It can be empty depending on observed_target_data and observed_source_data settings

Hrovatin committed:
Clarify tests

AdrianSosic committed:
Reuse parent method for integer casting

AdrianSosic committed:
Refactor GP fitting method

AdrianSosic committed:
Add temporary _task_parameter property to SearchSpace class

AdrianSosic committed:
Refactor transfer learning tests using parametrization/fixtures

AdrianSosic committed:
Update CHANGELOG.md

AdrianSosic commented on CHANGELOG.md:
Hi @Hrovatin 👋🏼  I've refactored the files and rebased everything onto main. From my part, the PR is ready to be merged. But now with these final changes, could you do me a little favor and just recreate the `before vs after` plots by running the new version against the version of the PRs base commit?

AVHopp commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -146,62 +146,60 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+        kernel = self.kernel_factory(context.searchspace, train_x, train_y)
+        covar_module = kernel.to_gpytorch(
             ard_num_dims=train_x.shape[-1] - context.n_task_dimensions,
-            active_dims=numerical_idxs,
```
One thing I figured out is that if `active_dims` is not provided, then this is NOT automatically set to the set of all indices but also stored as `None` in the `_buffers` of the object:

<img width="262" height="62" alt="Image" src="https://github.com/user-attachments/assets/b33e15be-d136-4cda-a0ef-68c676b2bd91" />

Maybe there is some check somewhere internally where this makes a difference, I'll see if I can figure something out.

AVHopp replied:
And it seems like this might have an influence on the setting of the `ScaleKernel` itself: https://github.com/cornellius-gp/gpytorch/blob/8433c0b86660d0f318e3b0b5175bcfc5f9967894/gpytorch/kernels/scale_kernel.py#L71 What happens when you simply add this argument again, do things then become reproducible?

AdrianSosic replied:
Could be. Luckily, things are reproducible locally, so I'm currently in the process of trimming down the code into a minimal example. Hope this brings some new insights

Scienfitz reviewed (approved):
lets do the final check referenced [here](https://github.com/emdgroup/baybe/pull/549#discussion_r2413523917) and merge if everything is alright

AdrianSosic committed:
Use parametrization instead of request

Hrovatin commented:
@AVHopp and @AdrianSosic the issue is indeed the handling of active dimensions in base kernel (as speculated [here](https://github.com/emdgroup/baybe/pull/549#discussion_r2409602277)).

Changing the active dims as in [this branch](https://github.com/emdgroup/baybe/tree/tl_benchmarking_investigation_activeDims ) ensures that in naive case the outcome matches exactly the main branch, while in non-naive case it still stays as was this branch.

These are equality analysis for Michalewicz (as that one was usually most obviously discrepant) - output True means equal:

```
tl-benchmarking-investigation-activeDims and main in naive True:
True
tl-benchmarking-investigation-activeDims and main in naive False:
False
tl-benchmarking-investigation-activeDims and tl-benchmarking-investigation in naive True:
False
tl-benchmarking-investigation-activeDims and tl-benchmarking-investigation in naive False:
True
main and tl-benchmarking-investigation in naive True:
False
main and tl-benchmarking-investigation in naive False:
False
```

Hrovatin committed:
Directly specify active_dims in kernel

AVHopp commented on baybe/surrogates/gaussian_process/core.py:
```diff
@@ -146,62 +146,62 @@ def _fit(self, train_x: Tensor, train_y: Tensor) -> None:
+        covar_module = kernel.to_gpytorch(
+            ard_num_dims=kernel_num_dims,
             batch_shape=batch_shape,
+            active_dims=tuple(range(kernel_num_dims)),
```
Ok, so how do we proceed here? @AdrianSosic seems to be working on a general minimal example for this issue (this is at least how I understand [this](https://github.com/emdgroup/baybe/pull/549#discussion_r2432044016)), so we basically only need to decide if we either wait for this investigation or include the change that was made in the branch that makes naive setting reproducible (btw I cannot access the link to that branch, maybe due to a rebasing?)

Hrovatin commented:
Note to myself: Have a look at https://github.com/meta-pytorch/botorch/issues/2739#issuecomment-3486075728

AdrianSosic committed:
Drop unnecessary arguments

The active_dims argument can now be dropped due to https://github.com/emdgroup/baybe/pull/671

Hrovatin commented:
There is a bug that was introduced between the test run on 9. 10. and 10. 11. in
`RuntimeError: Cannot create a task covariance matrix larger than the number of tasks`

AdrianSosic committed:
Pin botorch

https://github.com/meta-pytorch/botorch/issues/3085

AdrianSosic committed:
xfail tests on latest release instead of pinning version

Unfortunately, previous botorch version have an (unnecessary?)
hard pin for gpytorch on version 1.14, causing troubles with
other tests due to the following issue, which has only be fixed
in 1.14.1:
https://github.com/cornellius-gp/gpytorch/issues/2633

AdrianSosic committed:
Revert xfail

Does not solve the problem since there is still a failing example

AdrianSosic committed:
Set covariance matrix rank according to square root of number of tasks

AdrianSosic committed:
Switch to new botorch logic until better solution is available

AdrianSosic committed:
Fix type in botorch translation

AdrianSosic commented:
@copilot: Explain the reason for the CI failure

AdrianSosic commented:
Closed in favor of #743. In particular, we don't switch to `MultiTaskGP` because we do not want to adopt BoTorch's latest decisions on how to handle incoming task data (see `validate_task_values` argument). Staying with `SingleTaskGP` keeps full flexibility (also for implementing other transfer mechanisms that do not involve `IndexKernel`) and allows us to keep our GP class as a slim / general purpose skeleton that implements the core logic of a GP.
